### PR TITLE
[FLINK-11908][table] Port window classes into flink-api-java

### PR DESCRIPTION
**Pull request by hequn8128**
## What is the purpose of the change

This pull request ports window classes into flink-api-java module.

## Brief change log

  - Convert Tumble/Slide/Session/Over related classes to interface and rename the original class from XXX to XXXImpl. For example, add a TumbleWithSize interface and rename the original TumbleWithSize to TumbleWithSizeImpl. We can't port TumbleWithSize directly into the api-java module because ExpressionParser is used in it.
  - Deprecate `def window(window: Window): WindowedTable` and add `def window(window: GroupWindow): GroupWindowedTable`
  - Deprecate class WindowedTable and add GroupWindowedTable
  - Add test case, such as WindowValidationTest, to test failures for the creation of window. 

## Verifying this change

This change is already covered by existing Window IT/UT test cases.
Furthermore, WindowValidationTest has been add to test failures for the creation of window. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/GroupWindow.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * special type of groupBy. Just like groupBy, group windows allow to compute aggregates
+ * on groups of elements.
+ *
+ * <p>Infinite streaming tables can only be grouped into time or row intervals. Hence window
+ * grouping is required to apply aggregations on streaming tables.
+ *
+ * <p>For finite batch tables, group windows provide shortcuts for time-based groupBy.
+ *
+ * <p>Note: {@link Window} is temporally used as the father class of {@link GroupWindow} for the
+ * sake of compatibility. It will be removed later.
```
Comment (sunjincheng121) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/GroupWindow.java]: Here just a suggestion, how about we open a new PR for Deprecated the Window in release-1.8, since we should create a new RC2 for release 1.8. If we do not do that the Window will keep existing for almost half a year. And I'll create the Jira FLINK-11918, and link to release-1.8 vote mail thread, ask RM's options. If all of you do not agree, I'll close the JIRA, otherwise, we can open the new PR for Deprecated the window.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Tumble.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.lang.reflect.Constructor;
+
+/**
+ * Helper class for creating a tumbling window. Tumbling windows are consecutive, non-overlapping
+ * windows of a specified fixed length. For example, a tumbling window of 5 minutes size groups
+ * elements in 5 minutes intervals.
+ *
+ * <p>Java users should use the methods with parameters of String type, while Scala users should use
+ * the methods with parameters of Expression type.
```
Comment (sunjincheng121) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Tumble.java]: Same as above.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/table.scala [hunk 3]
```diff
@@ -1092,10 +1094,35 @@ class Table(
     * @param window window that specifies how elements are grouped.
     * @return A windowed table.
     */
+  @Deprecated
+  @deprecated(
+    "This method will be removed. Use table.window(window: GroupWindow) instead",
+    "1.9.0")
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/table.scala]: It depends on FLINK-11918

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 4]
```diff
@@ -333,7 +348,8 @@ class SlideWithSize(size: Expression) {
     * @param slide the slide of the window either as time or row-count interval.
     * @return a sliding window
     */
-  def every(slide: Expression): SlideWithSizeAndSlide = new SlideWithSizeAndSlide(size, slide)
+  def every(slide: Expression): SlideWithSizeAndSlide =
+    new SlideWithSizeAndSlideImpl(size, slide)
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/windows.scala]: Unnecessary change？

**Comment by sunjincheng121** [comment]
@flinkbot approve description
@flinkbot approve consensus

**Comment by hequn8128** [comment]
@sunjincheng121 Thanks a lot for your review. I have addressed all your comments.

@twalthr @sunjincheng121 I have also rebased to the deprecating window pr and delete the deprecated methods and classes in this one. Would be great if you can take a look. 

Best, Hequn

**Comment by hequn8128** [comment]
Update, rebase to the latest deprecating window [PR](https://github.com/apache/flink/pull/7985) to unblock review. The updates contain two commits. One is the PR of deprecating window, the other one is the PR of the current issue.

**Comment by hequn8128** [comment]
Rebase to the latest flink/master

**Comment by hequn8128** [comment]
@twalthr Sorry. I missed the important messages from you. Convert the ExpressionParser into an interface is a good solution. So we can port TumbleWithSize, TumbleWithSizeOnTime, etc. directly into api-java. This also performs a reflection step at just one location. 

Good idea! 
I will try to update the PR asap.

Best, Hequn

**Comment by hequn8128** [comment]
@twalthr @sunjincheng121 I have updated the PR. Would be great if you can take another look. 

Some changes are explained by the followings:
- To make `ExpressionParser.parseExpression()` available, the `ExpressionParser` contains two static methods: `parseExpression()` and `parseExpressionList()`.
- As static methods cannot be overridden, I added a `PlannerExpressionParser` interface. The `PlannerExpressionParser` contains two non-static expression parser methods and are overridden by `PlannerExpressionParserImpl`. So once we get a `PlannerExpressionParserImpl` instance through refection, we can call parseExpression() in `PlannerExpressionParserImpl`.
- Furthermore, I add a companion class `PlannerExpressionParserImpl`. This is because we can't new object directly.
- Both `ExpressionParser` and `PlannerExpressionParser` in api-java are marked as Internal class.

Best, 
Hequn

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitionedOrdered.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	 *
+	 * @param alias alias for this over window
+	 * @return the fully defined over window
+	 */
+	public OverWindow as(Expression alias) {
+		return new OverWindow(
+			alias,
+			partitionBy,
+			orderBy,
+			ExpressionParser.parseExpression("UNBOUNDED_RANGE"),
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitionedOrdered.java]: Don't introduce string constants at an arbitrary code location. Instead, either move it to the top of a class as a static final field or in this case use an already existing constant in `BuiltInFunctionDefinitions`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitioned.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Partially defined over window with partitioning.
+ */
+@PublicEvolving
+public final class OverWindowPartitioned {
+
+	/** Defines a partitioning of the input on one or more attributes. */
+	private final List<Expression> partitionBy;
+
+	public OverWindowPartitioned(List<Expression> partitionBy) {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindowPartitioned.java]: We can use a default visibility for intermediate API classes. Such that users can not accidentally use the wrong methods.

**Comment by twalthr** [comment]
@flinkbot approve all

**Comment by hequn8128** [comment]
@twalthr Thanks for your valuable suggestions. I will move on the next PR(convert Table to the interface).
